### PR TITLE
Setup infra rebuild for CaaS appliances

### PR DESCRIPTION
**Pull request by assumptionsandg**

### CODE REVIEW

File: roles/infra/tasks/rebuild.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        terraform_provision_state: "{{ terraform_state_pull.stdout | from_json }}"
+
+- name: Terraform look at the current state list
+  ansible.builtin.command: "{{ terraform_binary_path }} state list"
+  args:
+    chdir: "{{ terraform_project_path }}"
+  register: terraform_state_list
+
+- name: Taint all servers if this is a patch update
+  ansible.builtin.command: "{{ terraform_binary_path }} taint openstack_compute_instance_v2.cluster_server"
```
Comment (JohnGarbutt) [inline roles/infra/tasks/rebuild.yml:33]: I think this should be: {{ item }}

File: roles/infra/tasks/main.yml [hunk 2]
```diff
@@ -46,6 +46,10 @@
     content: "{{ _terraform_backend_config | to_json }}"
     dest: "{{ terraform_project_path }}/backend_config.json"
 
+- name: Rebuild infrastructure
+  include_tasks: rebuild.yml
+  when: terraform_rebuild_appliance == 'patch'
```
Comment (JohnGarbutt) [inline roles/infra/tasks/main.yml:51]: I tink you are missing `when not terraform_readonly` or maybe fail on it?

File: roles/infra/tasks/inventory_adopt.yml [hunk 3]
```diff
@@ -73,6 +73,7 @@
   wait_for:
     port: 22
     host: "{{ cluster_gateway_ip }}"
+    search_regex: OpenSSH
```
Comment (mkjpryor) [inline roles/infra/tasks/inventory_adopt.yml:76]: Can this be a separate PR, along with the same on the other `wait_for` please?

File: roles/infra/tasks/main.yml [hunk 4]
```diff
@@ -46,6 +46,12 @@
     content: "{{ _terraform_backend_config | to_json }}"
     dest: "{{ terraform_project_path }}/backend_config.json"
 
+- name: Set terraform to rebuild patched appliances
+  include_tasks: rebuild.yml
+  when:
+    - terraform_rebuild_appliance == 'patch'
```
Comment (mkjpryor) [inline roles/infra/tasks/main.yml:52]: Can we just make this a boolean please? I can't see why it would need to have more than yes/no semantics. Also, can this role doesn't really know the "appliance" terminology so maybe call this `terraform_rebuild_nodes`?

**Review by mkjpryor — CHANGES_REQUESTED** [review]
Just a couple of minor things. Otherwise looks good.

File: roles/infra/tasks/main.yml [hunk 5]
```diff
@@ -46,6 +46,12 @@
     content: "{{ _terraform_backend_config | to_json }}"
     dest: "{{ terraform_project_path }}/backend_config.json"
 
+- name: Set terraform to rebuild patched appliances
+  include_tasks: rebuild.yml
+  when:
+    - terraform_rebuild_nodes
```
Comment (mkjpryor) [inline roles/infra/tasks/main.yml:52]: We should set a role default that sets this to `false` so that we don't get undefined variable warnings when it isn't supplied, e.g. if the appliance doesn't support it yet.

**Review by mkjpryor — CHANGES_REQUESTED** [review]

File: roles/infra/tasks/main.yml [hunk 6]
```diff
@@ -46,6 +46,12 @@
     content: "{{ _terraform_backend_config | to_json }}"
     dest: "{{ terraform_project_path }}/backend_config.json"
 
+- name: Set terraform to rebuild patched appliances
+  include_tasks: rebuild.yml
+  when:
+    - terraform_rebuild_nodes | default(false)
```
Comment (mkjpryor) [inline roles/infra/tasks/main.yml:52]: @assumptionsandg 

I meant set a default in `roles/infra/defaults/main.yml` rather than this. Can we just change that? 🙂

**Review by mkjpryor — CHANGES_REQUESTED** [review]

File: roles/infra/tasks/main.yml (hunk 6 above)
Comment (assumptionsandg) [reply roles/infra/tasks/main.yml]: Done

**Review by JohnGarbutt — APPROVED** [review]
Nice, this looks good now.

**Review by mkjpryor — APPROVED** [review]
LGTM
